src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -6944,12 +6945,26 @@ static bool dependsOnTypeParameter(TypeParameterSymbol typeParameter1, TypeParam
                 fromExplicitCast: false,
                 diagnosticLocation: operandLocation);
 
+            // in the case of a lifted conversion, we assume that the call to the operator occurs only if the argument is not-null
+            if (!isLiftedConversion)
+            {
+                CheckDisallowedNullAssignment(operandType, parameterAnnotations, conversionOperand.Syntax.Location);
+            }
+
             // method parameter type -> method return type
             var methodReturnType = methodOpt.ReturnTypeWithAnnotations;
             operandType = GetLiftedReturnTypeIfNecessary(isLiftedConversion, methodReturnType, operandState);
-            if (!isLiftedConversion)
+            if (!isLiftedConversion || operandState.IsNotNull())
```

Comment:
Do we need the ` || operandState.IsNotNull()` part here?

Reply:
Hmm. Perhaps I should move the comment on line 6948 down here for clarity. It sort of applies to both of the checks.

Basically the `operandState.IsNotNull()` is necessary because in the case where we `isLiftedConversion` is true, we assume the operand must be not-null in order for the user-defined conversion operator to actually be called, and thus any postconditions on the return value should be respected.

A test which is affected by this condition is `UserDefinedConversion_UnconditionalParamAttributes_Lifted`. See `struct C` and `struct D`.

Reply:
Alternatively, we could make this part unconditional, and add something afterwards like `if (isLiftedConversion) operandType = operandType.WithState(operandType.State.Join(NullableFlowState.MaybeNull));`. But just it achieves the same behavior while doing a bit more work.